src/advanced_inv.cpp

```diff
@@ -761,49 +798,41 @@ bool advanced_inventory_pane::is_filtered( const item *it ) const
     return filtercache[ str ];
 }
 
-template <typename Container>
-static itemslice i_stacked( Container items )
+// roll our own, to handle moving stacks better
+typedef std::vector<std::list<item*>> itemstack;
+
+template <typename T>
+static itemstack i_stacked(T items)
 {
     //create a new container for our stacked items
-    itemslice islice;
-
-    //iterate through all items in the vector
-    for( auto &items_it : items ) {
-        if( items_it.count_by_charges() ) {
-            // Those exists as a single item all the item anyway
-            islice.push_back( std::make_pair( &items_it, 1 ) );
-            continue;
-        }
-        bool list_exists = false;
-
-        //iterate through stacked item lists
-        for( auto &elem : islice ) {
-            //check if the ID exists
-            item *first_item = elem.first;
-            if( first_item->type->id == items_it.type->id ) {
-                //we've found the list of items with the same type ID
-
-                if( first_item->stacks_with( items_it ) ) {
-                    //add it to the existing list
-                    elem.second++;
-                    list_exists = true;
+    itemstack stacks;
+    // used to recall index we stored `itype_id' item at in itemstack
+    std::unordered_map<itype_id, int> cache;
+    // iterate through and create stacks
+    for(auto &elem : items) {
+        const auto &id = elem.type->id;
+        auto iter = cache.find(id);
+        bool got_stacked = false;
+        // cache entry exists
+        if(iter != cache.end()) {
+            // check to see if it stacks with each item in a stack, not just front()
+            for(auto &it : stacks[iter->second]) {
+                if((got_stacked = it->stacks_with(elem))) {
+                    stacks[iter->second].push_back(&elem);
                     break;
                 }
             }
         }
-
-        if( !list_exists ) {
-            //insert the list into islice
-            islice.push_back( std::make_pair( &items_it, 1 ) );
+        // if the item didn't stack, add its index to cache and start the list
+        if(!got_stacked) {
+            cache[id] = stacks.size();
```

Comment:
There is still a very subtle bug. Items in this order on the ground: "can of A", "can of B", "can of A".

After the first item, the cache is {{ "can", 0 }}.
After the second item (which does not stack with the first one), it is {{ "can", 1 }}. (This line here overwrites the entry.)
Now the third item checks `stacks[cache["can"]] == stacks[1]`, which misses the already existing "can of A" stack at position 0. Why do you actually need to change that function at all?